### PR TITLE
Implement the logic in tpu estimator to support user provided stopping signals and save a checkpoint before stop.

### PR DESCRIPTION
**Pull request by ericzhangn**
Implement the logic in tpu estimator to support user provided
True/False stopping signals in the input data. When tpu estimator
encounters the user provided stopping signals with True value, it will save a
checkpoint automatically after finishing current iteration_per_loop and stops the
training. For users to use this feature, they will need add stopping signals with False
value in their real input data. And they will add some enough empty data after real input
data in the end which they know how to handle in their model code and have stopping 
signals with True value.

### CODE REVIEW

**Review by bfontain — COMMENTED** [review]
Looks good! Thanks for putting the changes behind the newly added TPUConfig parameter. I assume for TPU Embedding features, it is the caller's responsibility to provide empty batches for the padding features?

I can't approve right away since this changes the public API (via TPUConfig), I'm looping in the api team to give. Once they give the OK I can approve the request.

**Comment by ericzhangn** [comment]
> Looks good! Thanks for putting the changes behind the newly added TPUConfig parameter. I assume for TPU Embedding features, it is the caller's responsibility to provide empty batches for the padding features?
> 
> I can't approve right away since this changes the public API (via TPUConfig), I'm looping in the api team to give. Once they give the OK I can approve the request.

Thanks Bruce and your suggestions when we implemented this feature. Yes, for TPU embedding features, it is the caller's responsibility to provide empty batches for the padding features.

Sounds good. I will wait for the API team's further comments. Thank you for helping coordinate this.

**Comment by karmel** [comment]
For @tensorflow/api-owners 

We are concerned about the maintenance cost of this for all users. Is migrating to TF2, where this sort of customization is easier, an option? Or creating a subclass of the TPUConfig class?

**Comment by ericzhangn** [comment]
> For @tensorflow/api-owners
> 
> We are concerned about the maintenance cost of this for all users. Is migrating to TF2, where this sort of customization is easier, an option? Or creating a subclass of the TPUConfig class?

Thanks Karmel! Understand about the concern for related API change although it has some default value. I discussed Bruce and I think we might be able to make an even smaller change with majorly user customized implementation and does not need API change. Let me figure out some details and update the change.

**Comment by ericzhangn** [comment]
Update the code change based on code review comments and discussion with Bruce. Here the TPU estimator lets user to provide a customized TPU infeed outfeed hook class which can be used to add user specific logic. For example, we can implement this class to add end of stream handling and auto checkpointing.

File: tensorflow_estimator/python/estimator/tpu/tpu_config.py
Comment (fchollet) [inline tensorflow_estimator/python/estimator/tpu/tpu_config.py:132]: API owners notes: consider renaming this `experimental_feed_hook` (and `feed_hook` later).

**Comment by fchollet** [comment]
For @tensorflow/api-owners: we can merge this (with the name change), after LGTM from @bfontain.

**Review by fchollet — APPROVED** [review]
For @tensorflow/api-owners: we can merge this (with the name change), after LGTM from @bfontain.

File: tensorflow_estimator/python/estimator/tpu/tpu_config.py
Comment (ericzhangn) [reply tensorflow_estimator/python/estimator/tpu/tpu_config.py:132]: > For @tensorflow/api-owners: we can merge this (with the name change), after LGTM from @bfontain.
Great, thank you François!

**Review by bfontain — APPROVED** [review]
Looks good to me.

File: tensorflow_estimator/python/estimator/tpu/tpu_config.py
Comment (ericzhangn) [reply tensorflow_estimator/python/estimator/tpu/tpu_config.py]: Renaming is done.

**Comment by ericzhangn** [comment]
> Looks good to me.

Thanks Bruce!

**Pull request merged**
